buffer/src/test/java/io/netty/buffer/AbstractCompositeByteBufTest.java

```diff
@@ -1805,4 +1807,40 @@ public boolean process(byte value) throws Exception {
         assertArrayEquals(new byte[] {6, 5, 4, 3, 2, 1}, arrayDesc);
         buf.release();
     }
+
+    @Test
+    public void componentSliceFromFlattenedCompositeShouldHaveCorrectView() {
+        ByteBufAllocator alloc = new PooledByteBufAllocator();
+
+        // Create underlying buffer spacious enough for the test data.
+        ByteBuf buf = alloc.buffer(32).writeBytes("---01234".getBytes(US_ASCII));
+
+        CompositeByteBuf composite1 = alloc.compositeBuffer(8)
+                .addFlattenedComponents(true, buf);
+        assertThat(composite1.readCharSequence(3, US_ASCII).toString()).isEqualTo("---");
+
+        CompositeByteBuf composite2 = alloc.compositeBuffer(8)
+                .addFlattenedComponents(true, composite1);
+        assertThat(composite2.toString(US_ASCII)).isEqualTo("01234");
+
+        // Remove the last component from the cumulation, then add it back.
+        int tailComponentIndex = composite2.numComponents() - 1;
+        int tailStart = composite2.toByteIndex(tailComponentIndex);
+        ByteBuf component = composite2.componentSlice(tailComponentIndex);
+
+        assertThat(component.readableBytes()).isEqualTo(5);
+        assertThat(component.toString(US_ASCII)).isEqualTo("01234");
+
+        // Take the ownership of the component.
+        component.retain();
+
+        // Remove the component from the composite buf.
+        composite2.removeComponent(tailComponentIndex).setIndex(0, tailStart);
+
+        // Add it back to the composite byte buf.
+        composite2.addFlattenedComponents(true, component);
+
+        assertThat(composite2.toString(US_ASCII)).isEqualTo("01234");
```

Comment:
You need to release the buffers. This test is leaking memory, and it's failing the build.